File: backend/src/main/java/com/constructmanager/service/CategoryTeamService.java
package com.constructmanager.service;

import com.constructmanager.dto.*;
import com.constructmanager.entity.CategoryTeam;
import com.constructmanager.exception.ResourceNotFoundException;
import com.constructmanager.repository.CategoryRepository;
import com.constructmanager.repository.CategoryTeamRepository;
import com.constructmanager.repository.TeamRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.CacheManager;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
@Transactional(readOnly = true)
public class CategoryTeamService {
    
    @Autowired
    private CategoryTeamRepository categoryTeamRepository;
    
    @Autowired
    private CategoryRepository categoryRepository;
    
    @Autowired
    private TeamRepository teamRepository;
    
    @Autowired
    private CacheManager cacheManager;
    
    @Autowired
    private ProgressService progressService;
    
    /**
     * Get category teams by category
     */
    public List<CategoryTeam> getCategoryTeamsByCategory(Long categoryId) {
        return categoryTeamRepository.findByCategoryIdOrderByCreatedAtAsc(categoryId);
    }
    
    /**
     * Get category teams by team
     */
    public Page<CategoryTeam> getCategoryTeamsByTeam(Long teamId, Pageable pageable) {
        return categoryTeamRepository.findByTeamIdOrderByCreatedAtDesc(teamId, pageable);
    }
    
    /**
     * Get category team by category and team
     */
    public Optional<CategoryTeam> getCategoryTeam(Long categoryId, Long teamId) {
        return categoryTeamRepository.findByCategoryIdAndTeamId(categoryId, teamId);
    }
    
    /**
     * Create category team assignment
     */
    @Transactional
    public Optional<CategoryTeam> createCategoryTeam(Long categoryId, CategoryTeamCreateDTO dto) {
        return categoryRepository.findById(categoryId)
                .flatMap(category -> teamRepository.findById(dto.getTeamId())
                        .map(team -> {
                            CategoryTeam categoryTeam = new CategoryTeam();
                            categoryTeam.setCategory(category);
                            categoryTeam.setTeam(team);
                            categoryTeam.setStatus(dto.getStatus());
                            categoryTeam.setReceptionStatus(dto.getReceptionStatus());
                            categoryTeam.setPaymentStatus(dto.getPaymentStatus());
                            categoryTeam.setNotes(dto.getNotes());
                            
                            CategoryTeam savedCategoryTeam = categoryTeamRepository.save(categoryTeam);
                            
                            // Evict the categoryDetails cache for the unit associated with this category team
                            if (category.getUnit() != null) {
                                Long unitId = category.getUnit().getId();
                                cacheManager.getCache("categoryDetails").evict(unitId);
                            }
                            
                            return savedCategoryTeam;
                        }));
    }
    
    /**
     * Update category team assignment
     */
    @Transactional
    public CategoryTeam updateCategoryTeam(Long id, CategoryTeamUpdateDTO updateDTO) {
        CategoryTeam categoryTeam = categoryTeamRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("CategoryTeam not found with id: " + id));
        
        if (updateDTO.getStatus() != null) {
            categoryTeam.setStatus(updateDTO.getStatus());
        }
        
        if (updateDTO.getReceptionStatus() != null) {
            categoryTeam.setReceptionStatus(updateDTO.getReceptionStatus());
        }
        
        if (updateDTO.getPaymentStatus() != null) {
            categoryTeam.setPaymentStatus(updateDTO.getPaymentStatus());
        }
        
        if (updateDTO.getNotes() != null) {
            categoryTeam.setNotes(updateDTO.getNotes());
        }
        
        // Save the basic updates first
        CategoryTeam savedCategoryTeam = categoryTeamRepository.save(categoryTeam);
        
        // Evict the categoryDetails cache for the unit associated with this category team
        if (savedCategoryTeam.getCategory() != null && savedCategoryTeam.getCategory().getUnit() != null) {
            Long unitId = savedCategoryTeam.getCategory().getUnit().getId();
            cacheManager.getCache("categoryDetails").evict(unitId);
        }
        
        // If progress percentage is updated, use ProgressService to update and propagate changes
        if (updateDTO.getProgressPercentage() != null) {
            return progressService.updateCategoryTeamProgress(id, updateDTO.getProgressPercentage());
        }
        
        return savedCategoryTeam;
    }
    
    /**
     * Delete category team assignment
     */
    @Transactional
    public boolean deleteCategoryTeam(Long categoryTeamId) {
        return categoryTeamRepository.findById(categoryTeamId)
                .map(categoryTeam -> {
                    // Store the unitId before deleting the categoryTeam
                    Long unitId = null;
                    if (categoryTeam.getCategory() != null && categoryTeam.getCategory().getUnit() != null) {
                        unitId = categoryTeam.getCategory().getUnit().getId();
                    }
                    
                    categoryTeamRepository.delete(categoryTeam);
                    
                    // Evict the categoryDetails cache for the unit associated with this category team
                    if (unitId != null) {
                        cacheManager.getCache("categoryDetails").evict(unitId);
                    }
                    
                    return true;
                })
                .orElse(false);
    }
    
    /**
     * Get delayed tasks for alerts
     */
    public Page<CategoryTeam> getDelayedTasks(Long companyId, Pageable pageable) {
        return categoryTeamRepository.findDelayedTasks(companyId, pageable);
    }
    
    /**
     * Get tasks starting soon for alerts
     */
    public Page<CategoryTeam> getTasksStartingSoon(Long companyId, int daysAhead, Pageable pageable) {
        LocalDate today = LocalDate.now();
        LocalDate futureDate = today.plusDays(daysAhead);
        return categoryTeamRepository.findTasksStartingSoon(companyId, today, futureDate, pageable);
    }
    
    /**
     * Get tasks by status
     */
    public Page<CategoryTeam> getTasksByStatus(Long companyId, CategoryTeam.TaskStatus status, Pageable pageable) {
        return categoryTeamRepository.findTasksByStatus(companyId, status, pageable);
    }
    
    /**
     * Count tasks by status for dashboard
     */
    public Long countTasksByStatus(Long companyId, CategoryTeam.TaskStatus status) {
        return categoryTeamRepository.countTasksByStatus(companyId, status);
    }
    
    /**
     * Get tasks requiring payment
     */
    public Page<CategoryTeam> getTasksRequiringPayment(Long companyId, Pageable pageable) {
        return categoryTeamRepository.findTasksRequiringPayment(companyId, pageable);
    }
    
    /**
     * Get team workload
     */
    public Page<Object[]> getTeamWorkload(Long companyId, Pageable pageable) {
        return categoryTeamRepository.getTeamWorkload(companyId, pageable);
    }
}

File: src/lib/categoryTeamService.ts
// src/lib/categoryTeamService.ts
import { CategoryTeam } from '../types';
import { apiClient, PaginatedResponse } from './api';

// Define the backend's expected enum values as a union type
export type BackendTaskStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'DONE' | 'DELAYED';

export interface CategoryTeamCreateRequest {
  teamId: string;
  status?: BackendTaskStatus; // Use BackendTaskStatus here
  receptionStatus?: boolean;
  paymentStatus?: boolean;
  notes?: string;
}

export interface CategoryTeamUpdateRequest {
  status?: BackendTaskStatus; // Use BackendTaskStatus here
  receptionStatus?: boolean;
  paymentStatus?: boolean;
  notes?: string;
  progressPercentage?: number;
}

export class CategoryTeamService {
  static async getCategoryTeamsByCategory(categoryId: string): Promise<CategoryTeam[]> {
    // Correctly pass categoryId as a direct parameter
    return apiClient.get<CategoryTeam[]>('/category-teams', { categoryId: categoryId });
  }

  static async getCategoryTeamsByTeam(
    teamId: string,
    page = 0,
    size = 20
  ): Promise<PaginatedResponse<CategoryTeam>> {
    return apiClient.get<PaginatedResponse<CategoryTeam>>(`/category-teams/team/${teamId}`, {
      page: page.toString(),
      size: size.toString()
    });
  }

  static async getCategoryTeam(categoryId: string, teamId: string): Promise<CategoryTeam> {
    return apiClient.get<CategoryTeam>(`/category-teams/${categoryId}/${teamId}`);
  }

  static async createCategoryTeam(
    categoryId: string,
    data: CategoryTeamCreateRequest
  ): Promise<CategoryTeam> {
    return apiClient.post<CategoryTeam>(`/category-teams?categoryId=${categoryId}`, data);
  }

  static async updateCategoryTeam(
    id: string,
    data: CategoryTeamUpdateRequest
  ): Promise<CategoryTeam> {
    try {
      // Convert string ID to number for backend compatibility
      const numericId = parseInt(id, 10);
      if (isNaN(numericId)) {
        throw new Error(`Invalid category team ID: ${id}`);
      }
      console.log('Updating category team with ID:', numericId, 'and data:', data);
      return apiClient.put<CategoryTeam>(`/category-teams/${numericId}`, data);
    } catch (error) {
      console.error('Error in updateCategoryTeam:', error);
      throw error;
    }
  }

  static async deleteCategoryTeam(id: string): Promise<void> {
    return apiClient.delete<void>(`/category-teams/${id}`);
  }

  static async getDelayedTasks(companyId: string, page = 0, size = 20): Promise<PaginatedResponse<CategoryTeam>> {
    return apiClient.get<PaginatedResponse<CategoryTeam>>('/category-teams/delayed', {
      companyId: companyId,
      page: page.toString(),
      size: size.toString()
    });
  }

  static async getTasksStartingSoon(companyId: string, daysAhead = 7, page = 0, size = 20): Promise<PaginatedResponse<CategoryTeam>> {
    return apiClient.get<PaginatedResponse<CategoryTeam>>('/category-teams/starting-soon', {
      companyId: companyId,
      daysAhead: daysAhead.toString(),
      page: page.toString(),
      size: size.toString()
    });
  }

  static async getTasksByStatus(
    companyId: string,
    status: CategoryTeam['status'],
    page = 0,
    size = 20
  ): Promise<PaginatedResponse<CategoryTeam>> {
    return apiClient.get<PaginatedResponse<CategoryTeam>>(`/category-teams/status/${status}`, {
      companyId: companyId,
      page: page.toString(),
      size: size.toString()
    });
  }

  static async countTasksByStatus(companyId: string, status: CategoryTeam['status']): Promise<number> {
    return apiClient.get<number>(`/category-teams/count/${status}`, { companyId: companyId });
  }

  static async getTasksRequiringPayment(companyId: string, page = 0, size = 20): Promise<PaginatedResponse<CategoryTeam>> {
    return apiClient.get<PaginatedResponse<CategoryTeam>>('/category-teams/requiring-payment', {
      companyId: companyId,
      page: page.toString(),
      size: size.toString()
    });
  }

  static async getTeamWorkload(companyId: string, page = 0, size = 20): Promise<any> {
    return apiClient.get('/category-teams/workload', {
      companyId: companyId,
      page: page.toString(),
      size: size.toString()
    });
  }
}